pytorch_lightning/core/lightning.py

```diff
@@ -1718,6 +1718,24 @@ def _set_hparams(self, hp: Union[dict, Namespace, str]) -> None:
         else:
             self._hparams = hp
 
+    def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwargs):
+        """Saves the model in ONNX format
+
+        Args:
+            file_path: The path of the file the model should be saved to.
+            input_sample: A sample of an input tensor for tracing.
+            **kwargs: Will be passed to torch.onnx.export function.
+        """
+
+        if isinstance(input_sample, Tensor):
+            input_data = input_sample
+        elif self.example_input_array is not None:
+            input_data = self.example_input_array
+        else:
+            raise ValueError(f'input_sample and example_input_array tensors are both missing.')
+
+        torch.onnx.export(self, input_data, file_path, **kwargs)
```

Comment:
if the current `LightningModule` is a `ScriptModule` you need additional args, which should also be part of the signature I think (since they can be required in certain cases), whereas kwargs should only be for customisation

Reply:
Can we convert a `LightningModule` to a `ScriptModule` as of now? There seems to be an issue with this: https://github.com/PyTorchLightning/pytorch-lightning/issues/1951 . Also could you elaborate on the additional args? I'll be happy to incorporate the changes necessary :)

Reply:
You're right, currently this is not possible, but we certainly aim for it in the future and so we shouldn't block it then by this PR; from https://pytorch.org/docs/stable/onnx.html#torch.onnx.export it would be the `example_outputs` argument which is needed for script modules (although I don't understand why tbh)

Reply:
I was wondering if we could infer example_output given the input_data in the function itself. e.g:
```python
        self.eval()
        if 'example_outputs' not in kwargs:
            kwargs['example_outputs'] = self.forward(input_data)
        torch.onnx.export(self, input_data, file_path, **kwargs)
```
Would that work?

Reply:
I actually think this is best, if you change this to call `self(input_data)`, which implicitly calls `forward` together with some plain pytorch hook stuff :)

Reply:
@justusschock I have made the changes :]

Reply:
@justusschock I was wondering if it would be better to also provide another optional param to the user, to let him export the module using `script` or `trace`. i.e `mode='script'`. We could do the translation within the function.